pkg/chunk/schema_caching.go

```diff
@@ -90,24 +50,25 @@ func splitTimesByCacheability(from, through model.Time, cacheBefore model.Time)
 	return from, cacheBefore, cacheBefore, through
 }
 
-func mergeCacheableAndActiveQueries(cacheableQueries []IndexQuery, activeQueries []IndexQuery) []IndexQuery {
-	finalQueries := make([]IndexQuery, 0, len(cacheableQueries)+len(activeQueries))
+func (s *schemaCaching) setImmutability(from, through model.Time, f func(from, through model.Time) ([]IndexQuery, error)) ([]IndexQuery, error) {
+	queries, err := f(from, through)
+	if err != nil {
+		return nil, err
+	}
+
+	cacheBefore := model.TimeFromUnix(mtime.Now().Add(-s.cacheOlderThan).Unix())
 
-Outer:
-	for _, cq := range cacheableQueries {
-		for _, aq := range activeQueries {
-			// When deduping, the bucket values only influence TableName and HashValue
-			// and just checking those is enough.
-			if cq.TableName == aq.TableName && cq.HashValue == aq.HashValue {
-				continue Outer
-			}
+	// If the entire query is cacheable then cache it. We can't do partial query
+	// caching as we'd need bucket level info as to which bucket is immutable.
```

Comment:
This reasoning doesn't seem like a slam-dunk.  Why couldn't we require that all buckets before a certain time are immutable?

Reply:
Hrm, Its hard from a separation of concerns perspective. Right now, we are using a `cachingSchema` which wraps a normal `Schema`, buckets are schema internal, and looking at things at a bucket level would mean we move the caching logic into normal Schema itself.

Reply:
Maybe the comment could say something like "we don't know the start/end time of a bucket" ?
(If I understood your point)

Reply:
Done!